File: src/test/java/dev/hilla/generator/example/GenericsEndpoint.java
package dev.hilla.generator.example;

import dev.hilla.parser.annotations.Endpoint;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

@Endpoint
public class GenericsEndpoint {
  public List<GenericClass<String>> toListOfMyType(GenericClass<String> value) {
    return List.of(value);
  }

  public <T> List<T> toList(T[] value) {
    return Arrays.asList(value);
  }

  public <T, U> Map<T, GenericClass<U>> toMap(T key, GenericClass<U> value) {
    return Map.of(key, value);
  }

  public <T> GenericClass<GenericClass<T>> toNestedGenericClass(GenericClass<T> value) {
    return new GenericClass<>();
  }

  public <T, U, V>
      Map<GenericClass<List<T>>, Map<?, GenericClass<GenericClass<GenericClass<V>>>>> tooComplex(
          GenericClass<List<?>> p1, Map<U, GenericClass<GenericClass<GenericClass<?>>>> p2) {
    return null;
  }

  public static class GenericClass<T> {
    public T value;
  }
}

File: src/test/resources/dev/hilla/generator/example/GenericsEndpoint.ts
import client from '../../../../connect-client.default.js';
import type { GenericClass } from './GenericsEndpoint/GenericClass.js';
import type { EndpointRequestInit } from '@hilla/frontend';

const toList: { <T>(value: T[], init?: EndpointRequestInit): Promise<T[]> } = async (value, init) => {
    return client.call('dev.hilla.generator.example.GenericsEndpoint', 'toList', { value }, init);
};

const toListOfMyType: { (value: GenericClass<string>, init?: EndpointRequestInit): Promise<GenericClass<string>[]> } = async (value, init) => {
    return client.call('dev.hilla.generator.example.GenericsEndpoint', 'toListOfMyType', { value }, init);
};

const toMap: { <T, U>(key: T, value: GenericClass<U>, init?: EndpointRequestInit): Promise<Map<T, GenericClass<U>>> } = async (key, value, init) => {
    return client.call('dev.hilla.generator.example.GenericsEndpoint', 'toMap', { key, value }, init);
};

const toNestedGenericClass: { <T>(value: GenericClass<T>, init?: EndpointRequestInit): Promise<GenericClass<GenericClass<T>>> } = async (value, init) => {
    return client.call('dev.hilla.generator.example.GenericsEndpoint', 'toNestedGenericClass', { value }, init);
};

const GenericsEndpoint = {
    toList,
    toListOfMyType,
    toMap,
    toNestedGenericClass,
};

export default GenericsEndpoint;
